diff --git a/dts/upstream/src/riscv/allwinner/sunxi-d1s-t113.dtsi b/dts/upstream/src/riscv/allwinner/sunxi-d1s-t113.dtsi
index 82cc85acc..47986cb8b 100644
--- a/dts/upstream/src/riscv/allwinner/sunxi-d1s-t113.dtsi
+++ b/dts/upstream/src/riscv/allwinner/sunxi-d1s-t113.dtsi
@@ -90,6 +90,12 @@
 				function = "i2c3";
 			};
 
+			/omit-if-no-ref/
+			i2c3_pb_pins: i2c3-pb-pins {
+				pins = "PB6", "PB7";
+				function = "i2c3";
+			};
+
 			/omit-if-no-ref/
 			i2s1_pins: i2s1-pins {
 				pins = "PG12", "PG13";
@@ -174,6 +180,12 @@
 				function = "spi1";
 			};
 
+			/omit-if-no-ref/
+			uart1_pe_pins: uart1-pe-pins {
+				pins = "PE10", "PE11";
+				function = "uart1";
+			};
+
 			/omit-if-no-ref/
 			uart1_pg6_pins: uart1-pg6-pins {
 				pins = "PG6", "PG7";
