diff --git a/configs/infomat_dispkzn3_defconfig b/configs/infomat_dispkzn3_defconfig
index b91cb1043..566d04cb1 100644
--- a/configs/infomat_dispkzn3_defconfig
+++ b/configs/infomat_dispkzn3_defconfig
@@ -19,4 +19,8 @@ CONFIG_CMD_DIAG=y
 CONFIG_MISC=y
 CONFIG_RMII=y
 CONFIG_SUN8I_EMAC=y
-
+CONFIG_TOOLS_MKEFICAPSULE=n
+CONFIG_EFI_LOADER=n
+CONFIG_DM_I2C=y
+CONFIG_SYS_I2C_MVTWSI=y
+CONFIG_CMD_I2C=y
diff --git a/dts/upstream/src/arm/allwinner/sun8i-t113s-infomat-dispkzn3.dts b/dts/upstream/src/arm/allwinner/sun8i-t113s-infomat-dispkzn3.dts
index f811a6afd..011d37924 100644
--- a/dts/upstream/src/arm/allwinner/sun8i-t113s-infomat-dispkzn3.dts
+++ b/dts/upstream/src/arm/allwinner/sun8i-t113s-infomat-dispkzn3.dts
@@ -72,6 +72,12 @@
 	clock-frequency = <24000000>;
 };
 
+&wdt {
+	clocks = <&dcxo>;
+	clock-names = "hosc";
+	status = "okay";
+};
+
 &ehci0 {
 	status = "okay";
 };
@@ -89,21 +95,23 @@
 		compatible = "atmel,24c02";		/* actually it's a 24AA02E48 */
 		reg = <0x50>;
 		pagesize = <16>;
-		read-only;
 		vcc-supply = <&reg_vcc3v3>;
-
+/*
 		#address-cells = <1>;
 		#size-cells = <1>;
 
 		eth0_macaddress: macaddress@fa {
 			reg = <0xfa 0x06>;
 		};
+*/
 	};
 };
 
 &emac {
+/*
 	nvmem-cells = <&eth0_macaddress>;
 	nvmem-cell-names = "mac-address";
+*/
 	pinctrl-0 = <&rmii_pe_pins>;
 	pinctrl-names = "default";
 	phy-handle = <&ext_rmii_phy>;
@@ -123,11 +131,11 @@
 	pinctrl-0 = <&mmc0_pins>;
 	pinctrl-names = "default";
 	vmmc-supply = <&reg_vcc3v3>;
-        non-removable;
+	non-removable;
 	/*
-        cd-gpios = <&pio 5 6 GPIO_ACTIVE_LOW>;
+	cd-gpios = <&pio 5 6 GPIO_ACTIVE_LOW>;
 	disable-wp;
-        */
+	*/
 	bus-width = <4>;
 	status = "okay";
 };
