diff --git a/arch/arm/boot/dts/allwinner/sun8i-t113s-infomat-dispkzn3.dts b/arch/arm/boot/dts/allwinner/sun8i-t113s-infomat-dispkzn3.dts
index 42346bedf..898ce15c6 100644
--- a/arch/arm/boot/dts/allwinner/sun8i-t113s-infomat-dispkzn3.dts
+++ b/arch/arm/boot/dts/allwinner/sun8i-t113s-infomat-dispkzn3.dts
@@ -107,6 +107,8 @@ tcon0_out_lcd: endpoint@0 {
 };
 
 &pwm {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pwm7_pd_pin>;
 	status = "okay";
 };
 
diff --git a/arch/arm/configs/sunxi_defconfig b/arch/arm/configs/sunxi_defconfig
index f4b8d8f7d..42c8d7aee 100644
--- a/arch/arm/configs/sunxi_defconfig
+++ b/arch/arm/configs/sunxi_defconfig
@@ -157,6 +157,7 @@ CONFIG_IIO=y
 CONFIG_AXP20X_ADC=y
 CONFIG_PWM=y
 CONFIG_PWM_SUN4I=y
+CONFIG_PWM_SUN8I=y
 CONFIG_PHY_SUN4I_USB=y
 CONFIG_PHY_SUN9I_USB=y
 CONFIG_NVMEM_SUNXI_SID=y
diff --git a/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi b/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi
index ab52b3da6..e87035263 100644
--- a/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi
+++ b/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi
@@ -212,6 +212,145 @@ uart3_pb_pins: uart3-pb-pins {
 				pins = "PB6", "PB7";
 				function = "uart3";
 			};
+
+			/omit-if-no-ref/
+			pwm0_pb_pin: pwm0-pb-pin {
+				pins = "PB5";
+				function = "pwm0";
+			};
+
+			/omit-if-no-ref/
+			pwm0_pd_pin: pwm0-pd-pin {
+				pins = "PD16";
+				function = "pwm0";
+			};
+
+			/omit-if-no-ref/
+			pwm0_pg_pin: pwm0-pg-pin {
+				pins = "PG12";
+				function = "pwm0";
+			};
+
+			/omit-if-no-ref/
+			pwm1_pb_pin: pwm1-pb-pin {
+				pins = "PB6";
+				function = "pwm1";
+			};
+
+			/omit-if-no-ref/
+			pwm1_pd_pin: pwm1-pd-pin {
+				pins = "PD17";
+				function = "pwm1";
+			};
+
+			/omit-if-no-ref/
+			pwm1_pg_pin: pwm1-pg-pin {
+				pins = "PG6";
+				function = "pwm1";
+			};
+
+			/omit-if-no-ref/
+			pwm2_pd_pin: pwm2-pd-pin {
+				pins = "PD18";
+				function = "pwm2";
+			};
+
+			/omit-if-no-ref/
+			pwm2_pe_pin: pwm2-pe-pin {
+				pins = "PE8";
+				function = "pwm2";
+			};
+
+			/omit-if-no-ref/
+			pwm2_pg_pin: pwm2-pg-pin {
+				pins = "PG13";
+				function = "pwm2";
+			};
+
+			/omit-if-no-ref/
+			pwm3_pd_pin: pwm3-pd-pin {
+				pins = "PD19";
+				function = "pwm3";
+			};
+
+			/omit-if-no-ref/
+			pwm3_pe_pin: pwm3-pe-pin {
+				pins = "PE9";
+				function = "pwm3";
+			};
+
+			/omit-if-no-ref/
+			pwm3_pg_pin: pwm3-pg-pin {
+				pins = "PG10";
+				function = "pwm3";
+			};
+
+			/omit-if-no-ref/
+			pwm4_pd_pin: pwm4-pd-pin {
+				pins = "PD20";
+				function = "pwm4";
+			};
+
+			/omit-if-no-ref/
+			pwm4_pe_pin: pwm4-pe-pin {
+				pins = "PE10";
+				function = "pwm4";
+			};
+
+			/omit-if-no-ref/
+			pwm4_pg_pin: pwm4-pg-pin {
+				pins = "PG5";
+				function = "pwm4";
+			};
+
+			/omit-if-no-ref/
+			pwm5_pd_pin: pwm5-pd-pin {
+				pins = "PD21";
+				function = "pwm5";
+			};
+
+			/omit-if-no-ref/
+			pwm5_pf_pin: pwm5-pf-pin {
+				pins = "PF6";
+				function = "pwm5";
+			};
+
+			/omit-if-no-ref/
+			pwm5_pg_pin: pwm5-pg-pin {
+				pins = "PG4";
+				function = "pwm5";
+			};
+
+			/omit-if-no-ref/
+			pwm6_pd_pin: pwm6-pd-pin {
+				pins = "PD9";
+				function = "pwm6";
+			};
+
+			/omit-if-no-ref/
+			pwm6_pf_pin: pwm6-pf-pin {
+				pins = "PF4";
+				function = "pwm6";
+			};
+
+			/omit-if-no-ref/
+			pwm6_pg_pin: pwm6-pg-pin {
+				pins = "PG1";
+				function = "pwm6";
+			};
+
+			/omit-if-no-ref/
+			pwm7_pd_pin: pwm7-pd-pin {
+				pins = "PD22";
+				function = "pwm7";
+			};
+
+			/omit-if-no-ref/
+			pwm7_pg_pin: pwm7-pg-pin {
+				pins = "PG0";
+				function = "pwm7";
+			};
+
 		};
 
 		pwm: pwm@2000c00 {
