diff --git a/Documentation/devicetree/bindings/iommu/allwinner,sun50i-h6-iommu.yaml b/Documentation/devicetree/bindings/iommu/allwinner,sun50i-h6-iommu.yaml
index a8409db4a..83d6d925b 100644
--- a/Documentation/devicetree/bindings/iommu/allwinner,sun50i-h6-iommu.yaml
+++ b/Documentation/devicetree/bindings/iommu/allwinner,sun50i-h6-iommu.yaml
@@ -18,7 +18,9 @@ properties:
 
   compatible:
     oneOf:
-      - const: allwinner,sun50i-h6-iommu
+      - enum:
+          - allwinner,sun20i-d1-iommu
+          - allwinner,sun50i-h6-iommu
       - const: allwinner,sun50i-h616-iommu
       - items:
           - const: allwinner,sun55i-a523-iommu
@@ -42,7 +44,6 @@ required:
   - reg
   - interrupts
   - clocks
-  - resets
 
 additionalProperties: false
 
diff --git a/arch/arm/configs/sunxi_defconfig b/arch/arm/configs/sunxi_defconfig
index c3b885337..9cd96640d 100644
--- a/arch/arm/configs/sunxi_defconfig
+++ b/arch/arm/configs/sunxi_defconfig
@@ -199,7 +199,8 @@ CONFIG_LEDS_CLASS=y
 CONFIG_LEDS_CLASS_MULTICOLOR=y
 CONFIG_LEDS_USER=y
 CONFIG_LEDS_SUN50I_A100=y
-# CONFIG_IOMMU_SUPPORT is not set
+CONFIG_IOMMU_SUPPORT=y
+CONFIG_SUN50I_IOMMU=y
 CONFIG_IIO=y
 CONFIG_AXP20X_ADC=y
 CONFIG_PWM=y
diff --git a/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi b/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi
index b9fe257f9..9b880b708 100644
--- a/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi
+++ b/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi
@@ -425,6 +425,13 @@ ths: thermal-sensor@2009400 {
 			nvmem-cell-names = "calibration";
 			#thermal-sensor-cells = <0>;
 		};
+		iommu: iommu at 2010000 {
+			compatible = "allwinner,sun20i-d1-iommu";
+			reg = <0x2010000 0x10000>;
+			interrupts = <SOC_PERIPHERAL_IRQ(64) IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&ccu CLK_BUS_IOMMU>;
+			#iommu-cells = <1>;
+		};
 
 		dmic: dmic@2031000 {
 			compatible = "allwinner,sun20i-d1-dmic",
@@ -931,6 +938,7 @@ mixer0: mixer@5100000 {
 				 <&display_clocks CLK_MIXER0>;
 			clock-names = "bus", "mod";
 			resets = <&display_clocks RST_MIXER0>;
+			iommus = <&iommu 2>;
 
 			ports {
 				#address-cells = <1>;
@@ -953,6 +961,7 @@ mixer1: mixer@5200000 {
 				 <&display_clocks CLK_MIXER1>;
 			clock-names = "bus", "mod";
 			resets = <&display_clocks RST_MIXER1>;
+			iommus = <&iommu 2>;
 
 			ports {
 				#address-cells = <1>;
@@ -968,6 +977,17 @@ mixer1_out_tcon_top_mixer1: endpoint {
 			};
 		};
 
+		g2d: g2d@5410000 {
+			compatible = "allwinner,sunxi-g2d";
+			reg = <0x5410000 0x40000>;
+			interrupts = <SOC_PERIPHERAL_IRQ(89) IRQ_TYPE_LEVEL_HIGH>; /* 121 - 32 */
+			clocks = <&ccu CLK_BUS_G2D>, <&ccu CLK_G2D>, <&ccu CLK_MBUS_G2D>;
+			clock-names = "bus", "mod", "ram";
+			resets = <&ccu RST_BUS_G2D>;
+			iommus = <&iommu 3>; // G2D master 3 
+			status = "okay";
+		};
+
 		dsi: dsi@5450000 {
 			compatible = "allwinner,sun20i-d1-mipi-dsi",
 				     "allwinner,sun50i-a100-mipi-dsi";
diff --git a/drivers/iommu/sun50i-iommu.c b/drivers/iommu/sun50i-iommu.c
index 6306570d5..5ad622e0c 100644
--- a/drivers/iommu/sun50i-iommu.c
+++ b/drivers/iommu/sun50i-iommu.c
@@ -1033,7 +1033,8 @@ static int sun50i_iommu_probe(struct platform_device *pdev)
 		goto err_free_cache;
 	}
 
-	iommu->reset = devm_reset_control_get(&pdev->dev, NULL);
+	/* iommu->reset = devm_reset_control_get(&pdev->dev, NULL);*/
+	iommu->reset = devm_reset_control_get_optional(&pdev->dev, NULL);
 	if (IS_ERR(iommu->reset)) {
 		dev_err(&pdev->dev, "Couldn't get our reset line.\n");
 		ret = PTR_ERR(iommu->reset);
@@ -1071,6 +1072,7 @@ static int sun50i_iommu_probe(struct platform_device *pdev)
 static const struct of_device_id sun50i_iommu_dt[] = {
 	{ .compatible = "allwinner,sun50i-h6-iommu", },
 	{ .compatible = "allwinner,sun50i-h616-iommu", },
+	{ .compatible = "allwinner,sun20i-d1-iommu", },
 	{ /* sentinel */ },
 };
 MODULE_DEVICE_TABLE(of, sun50i_iommu_dt);
