diff --git a/arch/arm/boot/dts/allwinner/sun8i-t113s-infomat-dispkzn3.dts b/arch/arm/boot/dts/allwinner/sun8i-t113s-infomat-dispkzn3.dts
index fef626b31..8d884289e 100644
--- a/arch/arm/boot/dts/allwinner/sun8i-t113s-infomat-dispkzn3.dts
+++ b/arch/arm/boot/dts/allwinner/sun8i-t113s-infomat-dispkzn3.dts
@@ -106,6 +106,10 @@ tcon0_out_lcd: endpoint@0 {
 	};
 };
 
+&g2d {
+	status = "okay";
+};
+
 &pwm {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pwm7_pd_pin>;
diff --git a/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi b/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi
index 5419c6d22..d4eeb7a92 100644
--- a/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi
+++ b/arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi
@@ -978,6 +978,17 @@ mixer1_out_tcon_top_mixer1: endpoint {
 			};
 		};
 
+		g2d: g2d@5410000 {
+			compatible = "allwinner,sunxi-g2d";
+			reg = <0x5410000 0x40000>;
+			interrupts = <SOC_PERIPHERAL_IRQ(89) IRQ_TYPE_LEVEL_HIGH>; /* 121 - 32 */
+			clocks = <&ccu CLK_BUS_G2D>, <&ccu CLK_G2D>, <&ccu CLK_MBUS_G2D>;
+			clock-names = "bus", "mod", "ram";
+			resets = <&ccu RST_BUS_G2D>;
+			iommus = <&iommu 3>; // G2D master 3 
+			status = "disabled";
+		};
+
 		dsi: dsi@5450000 {
 			compatible = "allwinner,sun20i-d1-mipi-dsi",
 				     "allwinner,sun50i-a100-mipi-dsi";
